cpp/src/groupby/hash/groupby.cu

```diff
@@ -307,6 +387,33 @@ void compute_single_pass_aggs(table_view const& keys,
   }
 }
 
+void compute_multi_pass_aggs(std::vector<aggregation_request> const& requests,
+                             cudf::detail::result_cache* sparse_results,
+                             cudaStream_t stream)
+{
+  for (size_t i = 0; i < requests.size(); i++) {
+    auto const& agg_v  = requests[i].aggregations;
+    auto const& values = requests[i].values;
+
+    for (auto&& agg : agg_v) {
+      if (is_fixed_width(values.type()) and agg->kind == aggregation::MEAN) {
+        auto count_agg    = std::make_unique<aggregation>(aggregation::COUNT_VALID);
+        auto count_result = sparse_results->get_result(i, *count_agg);
+        auto sum_agg      = std::make_unique<aggregation>(aggregation::SUM);
+        auto sum_result   = sparse_results->get_result(i, *sum_agg);
+        auto result       = cudf::detail::binary_operation(
+          sum_result,
+          count_result,
+          binary_operator::DIV,
+          cudf::detail::target_type(values.type(), aggregation::MEAN),
+          rmm::mr::get_current_device_resource(),
+          stream);
+        sparse_results->add_result(i, *agg, std::move(result));
+      }
```

Comment:
The logic of computing a compound aggregation from it's simple components should be encapsulated in the `compound_aggregation` object. Something like a `finalize` virtual method that takes as input the results from the simple components and returns the compound result.

Reply:
That `finalize` part is still pending. If `finalize` is added to `compound_aggregation`, it creates unwanted include dependency to column_view and result cache. So, thinking about implementing as separate module or `compound_aggregation` itself as separate module (not in inheritance of aggregations, encapsulation of simple components is possible as you said).
Your suggestions?